[JSC] Add MulHigh

Constellation opened the pull request:
#### a4cc57da7cecacc29d071929c8941b27b906fba2
<pre>
[JSC] Add MulHigh
<a href="https://bugs.webkit.org/show_bug.cgi?id=290088">https://bugs.webkit.org/show_bug.cgi?id=290088</a>
<a href="https://rdar.apple.com/147473808">rdar://147473808</a>

Reviewed by Yijia Huang.

This patch adds MulHigh / UMulHigh B3 opcodes, which is extracting upper
extended bits from binary multiplication. (32bit * 32bit = 64bit, and
64bit&apos;s upper 32bit). This plays a critical role for Div / Mod&apos;s
strength reduction, so we would like to have them in IR explicitly.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::mulHigh32):
(JSC::MacroAssemblerARM64::mulHigh64):
(JSC::MacroAssemblerARM64::uMulHigh32):
(JSC::MacroAssemblerARM64::uMulHigh64):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::x86MulHigh32):
(JSC::MacroAssemblerX86_64::x86MulHigh64):
(JSC::MacroAssemblerX86_64::x86UMulHigh32):
(JSC::MacroAssemblerX86_64::x86UMulHigh64):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::mull_r):
(JSC::X86Assembler::mulq_r):
(JSC::X86Assembler::imull_r):
(JSC::X86Assembler::imulq_r):
* Source/JavaScriptCore/b3/B3Const32Value.cpp:
(JSC::B3::Const32Value::mulHighConstant const):
(JSC::B3::Const32Value::uMulHighConstant const):
* Source/JavaScriptCore/b3/B3Const32Value.h:
* Source/JavaScriptCore/b3/B3Const64Value.cpp:
(JSC::B3::Const64Value::mulHighConstant const):
(JSC::B3::Const64Value::uMulHighConstant const):
* Source/JavaScriptCore/b3/B3Const64Value.h:
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::mulHighConstant const):
(JSC::B3::Value::uMulHighConstant const):
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
(JSC::B3::Value::typeFor):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/air/AirInstInlines.h:
(JSC::B3::Air::isX86MulHighHelperValid):
(JSC::B3::Air::isX86MulHigh32Valid):
(JSC::B3::Air::isX86MulHigh64Valid):
(JSC::B3::Air::isX86UMulHigh32Valid):
(JSC::B3::Air::isX86UMulHigh64Valid):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_8.cpp:
(testMulHigh64):
(testMulHigh32):
(testUMulHigh64):
(testUMulHigh32):

Canonical link: <a href="https://commits.webkit.org/292466@main">https://commits.webkit.org/292466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a4323bd6bb5faf5b61514559de71d935da7000d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96157 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5732 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101222 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46682 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16066 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24204 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73308 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/30537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12042 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86859 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53645 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4620 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46001 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88830 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103247 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94778 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23224 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82347 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23475 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82880 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81723 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26335 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3762 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16585 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15470 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23187 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118255 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22846 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26326 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24587 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
```diff
@@ -309,6 +309,30 @@ arm64: MultiplySignExtend32 U:G:32, U:G:32, D:G:64
+    Tmp, Tmp, Tmp
+
+x86: X86MulHigh32 U:G:32, UD:G:32, D:G:32
+    Tmp, Tmp*, Tmp*
```
What's star in `Tmp*` again? It seems `Tmp` means the allocator can pick any register. But we must pick a specific, pre-defined physical register (e.g., eax, edx) for `Tmp*`.

Constellation replied:
This means that these registers have special constraint, so AirInstInlines.h's helper function will be invoked to validate them.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Add MulHigh
https://bugs.webkit.org/show_bug.cgi?id=290088
rdar://147473808

Reviewed by Yijia Huang.

This patch adds MulHigh / UMulHigh B3 opcodes, which is extracting upper
extended bits from binary multiplication. (32bit * 32bit = 64bit, and
64bit's upper 32bit). This plays a critical role for Div / Mod's
strength reduction, so we would like to have them in IR explicitly.

* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::mulHigh32):
(JSC::MacroAssemblerARM64::mulHigh64):
(JSC::MacroAssemblerARM64::uMulHigh32):
(JSC::MacroAssemblerARM64::uMulHigh64):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::x86MulHigh32):
(JSC::MacroAssemblerX86_64::x86MulHigh64):
(JSC::MacroAssemblerX86_64::x86UMulHigh32):
(JSC::MacroAssemblerX86_64::x86UMulHigh64):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::mull_r):
(JSC::X86Assembler::mulq_r):
(JSC::X86Assembler::imull_r):
(JSC::X86Assembler::imulq_r):
* Source/JavaScriptCore/b3/B3Const32Value.cpp:
(JSC::B3::Const32Value::mulHighConstant const):
(JSC::B3::Const32Value::uMulHighConstant const):
* Source/JavaScriptCore/b3/B3Const32Value.h:
* Source/JavaScriptCore/b3/B3Const64Value.cpp:
(JSC::B3::Const64Value::mulHighConstant const):
(JSC::B3::Const64Value::uMulHighConstant const):
* Source/JavaScriptCore/b3/B3Const64Value.h:
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::mulHighConstant const):
(JSC::B3::Value::uMulHighConstant const):
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
(JSC::B3::Value::typeFor):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/air/AirInstInlines.h:
(JSC::B3::Air::isX86MulHighHelperValid):
(JSC::B3::Air::isX86MulHigh32Valid):
(JSC::B3::Air::isX86MulHigh64Valid):
(JSC::B3::Air::isX86UMulHigh32Valid):
(JSC::B3::Air::isX86UMulHigh64Valid):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_8.cpp:
(testMulHigh64):
(testMulHigh32):
(testUMulHigh64):
(testUMulHigh32):

Canonical link: https://commits.webkit.org/292466@main
